### PR TITLE
Use package entry_points to list providers

### PR DESCRIPTION
**Pull request by fcollonval**
Remove notebook package import

### CODE REVIEW

**Comment by bollwyvl** [comment]
Excellent start! From the quick skim:

I'd either:

- use entrypoints:

> it's basically guaranteed to be installed if jupyter_server (and
therefore nbconvert) is, so wouldn't be a real novel dep, is faster, and
hasn't (and probably won't) changed its API

- always use importlib_*, set a bottom pin, and not try to use stdlib

> It's changed api a couple time, and the backport seems to not always be
in sync with stdlib, and conditional dependencies make me sad

**Comment by fcollonval** [comment]
Thanks for the suggestions @bollwyvl - I was not aware of API inconsistency for `importlib_*`. So I updated this one using `entrypoints`.

I set the version constraint as in nbconvert
https://github.com/jupyter/nbconvert/blob/6f1a56f63a9a8dc8ee5178aa6d918db589b6d439/setup.py#L222

**Comment by bollwyvl** [comment]
Yeah I ran into it over on the crazy entry point pr on jupyter_core. I
guess once 3.7 is eol, it'll all work itself out.

Other minor thing: I guess I'd be a tad more explicit when making the enum,
e.g. MANAGER.keys() or something.

One thing with entry_points... I like a "lazy as possible" approach, where
imports are deferred until the thing is actually used. So instead of
hoisting the provider class, have a simple function that does the import
and returns the instance. Then it's safe to put that function in the top
level module... So (please pardon whatever my phone does to this):

jupyter_pullrequests.provider_v1 =
    gitlab = jupyterlab_pull-requests:get_gitlab_manager [gitlab]

Points at:

def get_gitlab_manager(cfg):
    from .managers.gitlab import Gitlab
    return Gitlab(cfg)

This also prevents the rest of the module from being imported under certain
conditions... It's a bummer there are so many gotchas... But still kinda
unique among languages, so I'm glad we have it.

**Comment by fcollonval** [comment]
Thanks @bollwyvl for the review

Regarding the delayed import, this is already the case as the entry point is imported when `load` method is called. And this is done at the latest possible time in `setup_handlers` at line 213.

I can prove that easily as my first implementation was to call `load` when building `MANAGERS` dictionary. But that raises an import error:

```
jupyterlab_pullrequests/handlers.py:16: in <module>
    from .base import MANAGERS, CommentReply, NewComment, PRConfig
jupyterlab_pullrequests/base.py:11: in <module>
    MANAGERS[entry.name] = entry.load()
../../miniconda3/envs/pull-requests/lib/python3.9/site-packages/entrypoints.py:82: in load
    mod = import_module(self.module_name)
...
E   AttributeError: partially initialized module 'jupyterlab_pullrequests.managers.github' has no attribute 'GitHubManager' (most likely due to a circular import)
```

So I would prefer keeping the current implementation that avoid additional convolution.

> I clarified the Enum source.

**Comment by bollwyvl** [comment]
Right, load _usually_ is when import stuff happens, but not always... If
the entrypoints is in a deeper part of the module, it has to import the
parent modules to get there...

We ran into this on jupyter_server, as it was using pytest_tornasync in its
pytest plugin, but that dependency was gated behind the test extra. In that
case, we just took out the pytest11 entry_point altogether.

The problem arises when there are (missing) optional dependencies at the
top of a file where the entry point is defined (or any module it
imports)... Boom.

Indeed we should probably have a ci excursion that doesn't have the gitlab
extra installed to verify correct behavior under a "normal" GitHub-only
install.

**Comment by fcollonval** [comment]
Sounds like `jupyter_server` brought lots of trouble. And indeed your approach with the loader function will avoid troubles. I'll modify the PR accordingly.

File: jupyterlab_pullrequests/base.py [hunk 1]
```diff
@@ -1,8 +1,15 @@
 from typing import List, NamedTuple, Optional
 
+import entrypoints
 from traitlets import Enum, Unicode, default
 from traitlets.config import Configurable
 
+# Supported third-party services
+MANAGERS = {}
+
+for entry in entrypoints.get_group_all("pull_requests_manager"):
```
Comment (bollwyvl) [inline jupyterlab_pullrequests/base.py:10]: Nitpick, but the [specification](https://packaging.python.org/specifications/entry-points/#data-model) suggests:

```
new group should use names starting with a PyPI name owned by the consumer 
project, followed by .. Group names must be one or more groups of letters, 
numbers and underscores, separated by dots (regex ^\w+(\.\w+)*$).
```

Further, as we might be changing stuff in the future, having an API version baked in would probably be good as the "last line of defense" against big API changes, so perhaps `jupyterlab_pullrequests.manager_v1`.

Comment (fcollonval) [reply jupyterlab_pullrequests/base.py]: Thanks for pointing me to the specification. I updated the name with your suggestion.

**Comment by fcollonval** [comment]
Ahah

With your extensive awesome opensource contributions, I would be a fool to throw you the stone for that 😉

**Pull request merged**